### PR TITLE
Feature - implement reference count for ConsumerImpl

### PR DESCRIPTION
**Pull request by lovelle**
Add reference count for ConsumerImpl in order to track reused instances of a
consumer instance returned by `subscribe()` method call.
Having the reference of subscribed consumer instances offers the ability to not
close a consumer until the last corresponding `close()` is being called.

Modifications:

  - Add field on ConsumerBase to track references of consumer instances
    subscribed by the user.
  - Add checks on ConsumerImpl to know whether unsubscribe() or close() action
    should be performed regarding of reference count being zero value.
  - Increment reference count when a previous built consumer instance is being
    used by caller.

Future considerations:

When optimization #3312 is going to be made for other consumers implementation
aside from ConsumerImpl it should add refCount checks on close and unsubscribe
methods.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 1]
```diff
@@ -253,6 +253,10 @@ public UnAckedMessageTracker getUnAckedMessageTracker() {
 
     @Override
     public CompletableFuture<Void> unsubscribeAsync() {
+        if (!shouldTearDown()) {
```
Comment (sijie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java]: I don't think we need this for `unsubscribe`.

technically if any consumer in a subscription unsubscribe, the subscription is unsubscribed.

Comment (lovelle) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java]: Great! I was dubious about `unsubscribe`.
Thanks, removed! :+1:

**Comment by sijie** [comment]
run integration tests
run java8 tests

**Comment by jiazhai** [comment]
run integration tests

**Comment by lovelle** [comment]
run unit tests

**Comment by lovelle** [comment]
run java8 tests

**Comment by lovelle** [comment]
run java8 tests

**Comment by merlimat** [comment]
Merged in 2.3.1 at 
e190534

**Pull request merged**